
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a0/NumberSetinC.svg/600px-NumberSetinC.svg.png" alt="Julia Logo" width="1000px">

## Números Complejos
Los números imáginarios en Julia se pueden representar agregando la partícula `im` al final de un número:

In [1]:
1im

0 + 1im

Al hacer ésto, el número y sus sumandos se combierten al tipo complejo:

In [ ]:
3+2im

Las mismas operaciones que con los reales son posibles; suma, resta, multiplicación, división, exponenciación, etc.

In [ ]:
(-1 + 2im)^(1 + 1im)

In [ ]:
3(2 - 5im)^-1.0

¿Qué sucede cuando se aplica una operación sobre un número complejo y uno real?

In [1]:
2 + 2.0 + 2im

4.0 + 2.0im

In [ ]:
1+1im

**NOTA**: Puedes escribir un número complejo de distintas formas, pero es importante ser atentos a la notación de las divisiones: 

`3/4im == 3/(4*im) == -(3/4*im)`

### Funciones
Ya que los numeros complejos son bidimencionales existen varias funciones útiles para manejar numeros complejos:

In [ ]:
real(1 + 2im) # Obtener la parte real de un número

In [ ]:
imag(1 + 2im) # Obtener la parte imáginaria de un número

In [ ]:
conj(1 + 2im) # Obtener el conjugado complejo de un número


In [ ]:
(1+2im)'      # Conjugado complejo, también ;)

In [ ]:
abs(1 + 2im)  # Obtener la magnitud del número complejo

In [ ]:
abs2(1 + 2im) # Obtener la magnitud al cuadrado (Distancia manhattan) de un número complejo

In [ ]:
angle(1 + 2im)# Obtener el ángulo de un número complejo

Las funciones trigonométircas, exponenciales y lógaritmicas también funcionan con números complejos. Pruebalas:

In [2]:
sin(1 + 2im)

3.165778513216168 + 1.9596010414216063im

¿Cual es la raíz cuadrada de -1?

In [5]:
sqrt(complex(-1))

0.0 + 1.0im

## Racionales
Los números racionales se construyen con el operador `//`:

In [6]:
4//8

1//2

De ellos puedes obtener su denominador o numerador con las funciones `denominator()` y `numerator()` respectivamente:

In [8]:
println(denominator(2//3))
println(numerator(2//3))

3
2


Prueba operaciones que te habrían sido útiles en la primaria. ¿Cuánto son dos tercios más un quinto?

In [9]:
2//3 + 1//5

13//15

Prueba comparar números racionales.

In [10]:
1//2 == 2//4

true

## Irracionales
Los números irracionales se definen en base a funciones.
Julia contiene por default algunos números irracionales:
```
π        3.14159265358979323846  pi
e        2.71828182845904523536  exp(big(1))
γ        0.57721566490153286061  euler
catalan  0.91596559417721901505  catalan
φ        1.61803398874989484820  (1+sqrt(big(5)))/2
```
https://github.com/JuliaLang/julia/blob/master/base/irrationals.jl

## Booleanos
El álgebra booleana, similar al algebra elementaria, trata solamente con dos valores (cierto y falso, uno y cero, arriba y abajo, etc.) Es muy útil para computo y lógica propocicional. Los booleanos en Julia se representan con dos valores: `true` y `false`. En programación los usarás comunmente en conjunto con los operadores `&&`, `||`, `!` (and, or, not) para hacer comparaciones lógicas:

In [11]:
true && true

true

In [1]:
false || !false

true

## Caractéres
Los caractéres en Julia se representan con el encoding [Unicode](https://unicode-table.com/en/) . Se escriben entre apostrofes de la siguiente manera:

In [12]:
`a`

`a`

In [13]:
'❤'

'❤': Unicode U+2764 (category So: Symbol, other)

Los caractéres son representados con números también. Ésto significa que puedes convertir entre distintos tipos:

In [14]:
Int('x')

120

120 es el número decimal con el que se representa en unicode la letra 'x'. ¿Cual es el número en binario que representa la primer letra de tu nombre?

In [16]:
Int('k')

107

Una forma de escribir unicode que no esté en tu teclado es con la siguiente notación:

In [17]:
'\u7A'

'z': ASCII/Unicode U+007a (category Ll: Letter, lowercase)

Cómo puedes ver, toma un número en hexadecimal y es equivalente a escribir `'\x78'`

In [18]:
'\x78'

'x': ASCII/Unicode U+0078 (category Ll: Letter, lowercase)

Algunas comparaciones son posibles entre caractéres. Dos operaciones artiméticas también lo son. ¿Cuáles son?

In [ ]:
'x' == 'x'

## Cadenas
Las cadenas de téxto son como el nombre lo indica, un encadenamiento de caractéres. En Julia se pueden escribir de dos formas; entre comillas para una linea:

In [ ]:
"Ésta es una cadena de téxto åø´¨æ≤≥ππµçæ≥¬πç©£¢º“¡¨ˆ"

o entre tres pares comillas seguidas para texto con saltos de linea. Támbién es útil escribir tus cadenas con tres comillas si contienen en sí comillas: 

In [20]:
""" Hola "mundo" 
ya llegué"""

" Hola mundo \nya llegué"

En la siguiente celda asignaremos el valor de una cadena a la variable `str`:

In [2]:
str = """Hola "mundo" 
ya llegué"""

"Hola \"mundo\" \nya llegué"

Ya que ésta variable contiene ahora el encadenamiento de varios caractéres podemos acceder a cada uno de los elementos de la variable con la siguiente notación:

In [ ]:
str[1]

In [ ]:
str[2]

In [ ]:
str[end]

In [ ]:
str[end-1]

In [5]:
str[1:4]
str[1:2]

"Ho"

Pero obtendremos un error al intentar acceder fuera de los límites de la cadena:

In [23]:
str[end+1]

LoadError: UnicodeError: invalid character index

In [ ]:
str[0]

### Concatenación
Concatenar es juntar dos cadenas y ee lleva a cabo con el operador `*`:

In [11]:
"Hola "*"Mundo"
println("HOLA\n"^6)

HOLA
HOLA
HOLA
HOLA
HOLA
HOLA



### Interpolación
Dentro de una cadena puedes introducir el valor de una variable de la siguiente manera:

In [25]:
"""Esa es mi cadena: "$str" """

"Esa es mi cadena: \"Hola \"mundo\" \nya llegué\" "

In [26]:
"1+2=$(1+2)"

"1+2=3"

### Funciones para cadenas
Éstas son algunas funciones útiles para el manejo de cadenas:

In [27]:
search("Hello, world.", 'o')

5

In [28]:
contains("Hello, world.", "world")

true

In [29]:
repeat(">.<", 10)

">.<>.<>.<>.<>.<>.<>.<>.<>.<>.<"

In [30]:
length("Hello, world.")

13

### Regex
Para quién esté interesado, Julia acepta regex:

In [31]:
r"^\s*(?:#|$)"

r"^\s*(?:#|$)"

In [32]:
match(r"^\s*(?:#|$)", "# a comment")

RegexMatch("#")